### DISCUSSION TITLE
SetUserAgentOverrideAsync

### DISCUSSION THREAD

**Discussion by wangf205x** [cefsharp/CefSharp#3485] [Q&A]
I am trying to set user agent.But the following code doesn't work.
					
					
code:
				
using (CefSharp.DevTools.DevToolsClient dtc = this.browser.GetDevToolsClient())
{
    IList<UserAgentBrandVersion> brandsList = new List<UserAgentBrandVersion>();
    UserAgentBrandVersion uab = new UserAgentBrandVersion();
    uab.Brand = "Google Chrome";
    uab.Version = "89";
    brandsList.Add(uab);

    UserAgentBrandVersion uab2 = new UserAgentBrandVersion();
    uab2.Brand = "Chromium";
    uab2.Version = "89";
    brandsList.Add(uab2);

    UserAgentMetadata ua = new UserAgentMetadata();
    ua.Brands = brandsList;
    ua.Architecture = "";
    ua.Model = "Nexus 7";
    ua.Platform = "Android";
    ua.PlatformVersion = "6.0.1";
    ua.FullVersion = "89.0.4389.114";
    ua.Mobile = true;

    dtc.Emulation.SetUserAgentOverrideAsync("Mozilla/5.0 (Linux; Android 6.0.1; Nexus 7 Build/MOB30X) AppleWebKit/5(KHTML,likeGeckoChrome/89.0.4389.114Safari/537.36",null,null,ua);
}

Os: windows 10
Cefsharp version: 89.0.4389.114
Platform: winforms
Cpu: x86

**Comment by campersau** [comment]
If you `await` the `SetUserAgentOverrideAsync` you will see that currently an exception is thrown because it looks like `IList`s of `DevToolsDomainEntityBase` aren't currently correctly serialized and need some special handling in `ToDictionary`:
https://github.com/cefsharp/CefSharp/blob/94fb45766de8e0b809bc9ef9195c4e0a9259919f/CefSharp/DevTools/DevToolsDomainEntityBase.cs#L131-L234

So currently it only works when setting `UserAgentMetadata.Brands` to an empty list. (Because it required and can't be `null` which might be wrong because here it is declared as `optional` https://github.com/ChromeDevTools/devtools-protocol/blob/7dd7cbb87cfb4f271d6a318364eea425f808ef15/json/browser_protocol.json#L7141.) Or leaving the `UserAgentMetadata` out completly and only set the `userAgent` string.

Complete Testcase with OffScreen Browser:
```c#
[Fact]
public async Task CanSetUserAgentOverride()
{
    using (var browser = new ChromiumWebBrowser("about:blank"))
    {
        await browser.LoadPageAsync();

        using (var devToolsClient = browser.GetDevToolsClient())
        {
            var brandsList = new List<UserAgentBrandVersion>();
            var uab = new UserAgentBrandVersion();
            uab.Brand = "Google Chrome";
            uab.Version = "89";
            brandsList.Add(uab); // comment this out to make it work

            var uab2 = new UserAgentBrandVersion();
            uab2.Brand = "Chromium";
            uab2.Version = "89";
            brandsList.Add(uab2); // comment this out to make it work

            var ua = new UserAgentMetadata();
            ua.Brands = brandsList;
            ua.Architecture = "";
            ua.Model = "Nexus 7";
            ua.Platform = "Android";
            ua.PlatformVersion = "6.0.1";
            ua.FullVersion = "89.0.4389.114";
            ua.Mobile = true;

            await devToolsClient.Emulation.SetUserAgentOverrideAsync("Mozilla/5.0 (Linux; Android 6.0.1; Nexus 7 Build/MOB30X) AppleWebKit/5(KHTML,likeGeckoChrome/89.0.4389.114Safari/537.36", null, null, ua);
        }

        await browser.LoadPageAsync("https://httpbin.org/user-agent");

        var html = await browser.GetTextAsync();
        Assert.Contains("Mozilla/5.0 (Linux; Android 6.0.1; Nexus 7 Build/MOB30X) AppleWebKit/5(KHTML,likeGeckoChrome/89.0.4389.114Safari/537.36", html);
    }
}
```

**Reply by wangf205x** [reply]
it still not work,only set the userAgent string
![image](https://user-images.githubusercontent.com/3354655/114065718-8d396a80-98cd-11eb-84c9-63202223513d.png)
![image](https://user-images.githubusercontent.com/3354655/114065820-a6421b80-98cd-11eb-8e63-2bd2e8c43913.png)

**Reply by amaitland** [reply]
Confirm your JavaScript works in DevTools when run using chrome.

**Reply by wangf205x** [reply]
if set userAgentData is successful,JavaScript will work

avigator.userAgentData.getHighEntropyValues(["architecture", "model", "platform", "platformVersion", "uaFullVersion"]).then(ua = >{
	console.log(ua)
});

https://web.dev/user-agent-client-hints/

**Reply by amaitland** [reply]
If the `userAgent` is being set correctly then it's working from a `CefSharp DevTools` point of view.

> https://web.dev/user-agent-client-hints/

~~https://web.dev/user-agent-client-hints/#user-agent-client-hints-from-chromium-84 suggests this is still experimental, in which version was it enabled by default? It may be that `CEF` requires code changes to support this.~~

**Reply by amaitland** [reply]
> So currently it only works when setting `UserAgentMetadata.Brands` to an empty list. (Because it required and can't be `null` which might be wrong because here it is declared as `optional` https://github.com/ChromeDevTools/devtools-protocol/blob/7dd7cbb87cfb4f271d6a318364eea425f808ef15/json/browser_protocol.json#L7141.)

The brands property doesn't have an `"optional": true` when I look at the `JSON` for version `89.0.4389.114`. So I think it's likely they've made it optional at a later point.

If I look at an older commit say https://github.com/ChromeDevTools/devtools-protocol/blob/769185f1ea71b8507e72ae245ecaaa3754c28c57/json/browser_protocol.json#L6958 then I don't see an optional value, so I'd say they made it optional in a newer version.

> if set userAgentData is successful,JavaScript will work

I've merged #3487, you can grab a `CI` build from https://www.myget.org/gallery/cefsharp and test it out. You'll need `90.2.30-CI4006`.

**Reply by wangf205x** [reply]
Thank you very much
![image](https://user-images.githubusercontent.com/3354655/114131016-53e81580-9934-11eb-873c-3b161e56e8d4.png)
